### PR TITLE
Add error messages to IPVS module

### LINKED ISSUE #89597: Make IPVS errors more verbose
So while digging into https://github.com/kubernetes/kubernetes/issues/89520 I've faced that the IPVS code isn't as verbose as the rest of the code, turning pretty hard to debug and identify where an error is ocurring.

The ``pkg/utils/ipvs`` package could contain a little bit more klog.Errorf in the `` if err !=nil as an example of a missing part here: https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/util/ipvs/ipvs_linux.go#L184-L186

/sig network
/priority important-long-term
/good-first-issue

/cc @aojea

### PR DESCRIPTION
**Pull request by carlosedp**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

This PR improves error messages on IPVS module.

**Which issue(s) this PR fixes**:

Improve IPVS module by adding error messages on convertion functions.

Fixes #89597.

Signed-off-by: Carlos de Paula <me@carlosedp.com>

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by rikatz** [comment]
@carlosedp thank you for the PR.

I think maybe you forgot also this: https://github.com/kubernetes/kubernetes/blob/518be88e7274942089abbfb3f5678fc6625d4e4b/pkg/util/ipvs/ipvs_linux.go#L109

**Review by rikatz — CHANGES_REQUESTED** [review]
In https://github.com/kubernetes/kubernetes/blob/518be88e7274942089abbfb3f5678fc6625d4e4b/pkg/util/ipvs/ipvs_linux.go#L109 also needs an Error Message

**Comment by carlosedp** [comment]
Oops, my bad... done.

**Review by andrewsykim — COMMENTED** [review]
We already log errors in the places where the ipvs util methods are called from the proxier. Wondering why those logs are not sufficient?

**Comment by rikatz** [comment]
@andrewsykim I've raised this as an issue because of https://github.com/kubernetes/kubernetes/issues/89520

It wasn't clear where the error was occurring, it just occurs in the outer functions but was pretty hard to identify in which part of the IPVS library (at least for me).

**Comment by andrewsykim** [comment]
I see, thanks for the context. I have concerns about logging the same error multiple times and making the logs harder to read. I wonder if instead we should be wrapping the errors with better descriptions so it's easier to find them. What do you think?

**Comment by rikatz** [comment]
sounds good to me :)

**Comment by rikatz** [comment]
@carlosedp as per the idea of Andrew, can you change the log errors and instead wrap them with better descriptions instead?

@andrewsykim just to confir, the idea is to "concatenate" the returned error, right?

**Comment by SataQiu** [comment]
+1 for @andrewsykim 
Wrapping the errors with descriptions is better :)

File: pkg/util/ipvs/ipvs_linux.go [hunk 1]
```diff
@@ -59,6 +59,7 @@ func New(exec utilexec.Interface) Interface {
 func (runner *runner) AddVirtualServer(vs *VirtualServer) error {
 	svc, err := toIPVSService(vs)
 	if err != nil {
+		klog.Errorf("Could not convert IPVS while adding server, error: %v", err)
 		return err
```
Comment (aojea) [inline pkg/util/ipvs/ipvs_linux.go:63]:
```suggestion
		return errors.Wrap(err, "Could not convert IPVS while adding server")
```

**Comment by carlosedp** [comment]
That's much better. Thanks for the suggestion.

**Comment by aojea** [comment]
@carlosedp I thought it was already importing `"github.com/pkg/errors"` , 
you have to import it and run `hack/update-bazel.sh`

**Comment by carlosedp** [comment]
Should I replace Golang's "error" with "github.com/pkg/errors" or import it naming it something else?

**Comment by aojea** [comment]
> Should I replace Golang's "error" with "github.com/pkg/errors" or import it naming it something else?

 replace Golang's "errors" with "github.com/pkg/errors" IIRC

**Comment by aojea** [comment]
/lgtm 
thanks

**Comment by carlosedp** [comment]
Any news on this?

**Comment by carlosedp** [comment]
Ping @andrewsykim :)

File: pkg/util/ipvs/ipvs_linux.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ func New(exec utilexec.Interface) Interface {
 func (runner *runner) AddVirtualServer(vs *VirtualServer) error {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS while adding server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:62]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 3]
```diff
@@ -70,7 +70,7 @@ func (runner *runner) AddVirtualServer(vs *VirtualServer) error {
 func (runner *runner) UpdateVirtualServer(vs *VirtualServer) error {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS while updating server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:73]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 4]
```diff
@@ -81,7 +81,7 @@ func (runner *runner) UpdateVirtualServer(vs *VirtualServer) error {
 func (runner *runner) DeleteVirtualServer(vs *VirtualServer) error {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS while deleting server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:84]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 5]
```diff
@@ -92,18 +92,18 @@ func (runner *runner) DeleteVirtualServer(vs *VirtualServer) error {
 func (runner *runner) GetVirtualServer(vs *VirtualServer) (*VirtualServer, error) {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Could not convert IPVS while getting server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:95]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 6]
```diff
@@ -92,18 +92,18 @@ func (runner *runner) DeleteVirtualServer(vs *VirtualServer) error {
...
 	}
 	runner.mu.Lock()
 	ipvsSvc, err := runner.ipvsHandle.GetService(svc)
 	runner.mu.Unlock()
 
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Could not get IPVS service")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:102]: nit: start errors with lower case

File: pkg/util/ipvs/ipvs_linux.go [hunk 7]
```diff
@@ -92,18 +92,18 @@ func (runner *runner) DeleteVirtualServer(vs *VirtualServer) error {
...
 	}
 	vServ, err := toVirtualServer(ipvsSvc)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Could not convert IPVS service to virtual server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:106]: "could not convert IPVS service to local virtual server"

File: pkg/util/ipvs/ipvs_linux.go [hunk 8]
```diff
@@ -114,13 +114,13 @@ func (runner *runner) GetVirtualServers() ([]*VirtualServer, error) {
 	ipvsSvcs, err := runner.ipvsHandle.GetServices()
 	runner.mu.Unlock()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Could not get IPVS services")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:117]: nit: start with lower case

File: pkg/util/ipvs/ipvs_linux.go [hunk 9]
```diff
@@ -114,13 +114,13 @@ func (runner *runner) GetVirtualServers() ([]*VirtualServer, error) {
...
 	}
 	vss := make([]*VirtualServer, 0)
 	for _, ipvsSvc := range ipvsSvcs {
 		vs, err := toVirtualServer(ipvsSvc)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, "Could not convert IPVS service to virtual server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:123]: "could not convert IPVS service to local virtual server"

File: pkg/util/ipvs/ipvs_linux.go [hunk 10]
```diff
@@ -138,11 +138,11 @@ func (runner *runner) Flush() error {
 func (runner *runner) AddRealServer(vs *VirtualServer, rs *RealServer) error {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS while adding real server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:141]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 11]
```diff
@@ -138,11 +138,11 @@ func (runner *runner) Flush() error {
...
 	}
 	dst, err := toIPVSDestination(rs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS service to real destination")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:145]: "could not convert local real server to IPVS destination"

File: pkg/util/ipvs/ipvs_linux.go [hunk 12]
```diff
@@ -153,11 +153,11 @@ func (runner *runner) AddRealServer(vs *VirtualServer, rs *RealServer) error {
 func (runner *runner) DeleteRealServer(vs *VirtualServer, rs *RealServer) error {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS while deleting real server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:156]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 13]
```diff
@@ -153,11 +153,11 @@ func (runner *runner) AddRealServer(vs *VirtualServer, rs *RealServer) error {
...
 	}
 	dst, err := toIPVSDestination(rs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS service to real destination")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:160]: "could not convert local real server to IPVS destination"

File: pkg/util/ipvs/ipvs_linux.go [hunk 14]
```diff
@@ -167,11 +167,11 @@ func (runner *runner) DeleteRealServer(vs *VirtualServer, rs *RealServer) error
 func (runner *runner) UpdateRealServer(vs *VirtualServer, rs *RealServer) error {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS while updating real server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:170]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 15]
```diff
@@ -167,11 +167,11 @@ func (runner *runner) DeleteRealServer(vs *VirtualServer, rs *RealServer) error
...
 	}
 	dst, err := toIPVSDestination(rs)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "Could not convert IPVS service to IPVS destination")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:174]: "could not convert local real server to IPVS destination"

File: pkg/util/ipvs/ipvs_linux.go [hunk 16]
```diff
@@ -182,20 +182,20 @@ func (runner *runner) UpdateRealServer(vs *VirtualServer, rs *RealServer) error
 func (runner *runner) GetRealServers(vs *VirtualServer) ([]*RealServer, error) {
 	svc, err := toIPVSService(vs)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Could not convert IPVS while getting real server")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:185]: "could not convert local virtual server to IPVS service"

File: pkg/util/ipvs/ipvs_linux.go [hunk 17]
```diff
@@ -182,20 +182,20 @@ func (runner *runner) UpdateRealServer(vs *VirtualServer, rs *RealServer) error
...
 	}
 	runner.mu.Lock()
 	dsts, err := runner.ipvsHandle.GetDestinations(svc)
 	runner.mu.Unlock()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Could not get IPVS destination for service")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:191]: nit: start with lower case

File: pkg/util/ipvs/ipvs_linux.go [hunk 18]
```diff
@@ -182,20 +182,20 @@ func (runner *runner) UpdateRealServer(vs *VirtualServer, rs *RealServer) error
...
 	}
 	rss := make([]*RealServer, 0)
 	for _, dst := range dsts {
 		dst, err := toRealServer(dst)
 		// TODO: aggregate errors?
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, "Could not convert IPVS destination to real destination")
```
Comment (andrewsykim) [inline pkg/util/ipvs/ipvs_linux.go:198]: "could not convert IPVS destination to local real server"

**Review by andrewsykim — COMMENTED** [review]
Sorry for the delay, had mostly nits around formatting.

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

Thank you @carlosedp

**Comment by carlosedp** [comment]
All good now!

**Comment by rikatz** [comment]
@carlosedp tks for taking this one :)

**Pull request merged**